finality-aleph/src/data_io/data_store.rs

```diff
@@ -300,18 +302,46 @@ where
 
         let now = time::SystemTime::now();
         for (proposal, first_occurrence) in proposals_with_timestamps {
-            if !self.bump_proposal(&proposal) {
-                // `bump_proposal` returns false if the bump didn't make the proposal available, meaning that it is still pending
-                if let Ok(time_waiting) = now.duration_since(first_occurrence) {
-                    if time_waiting >= self.config.request_block_after {
-                        let block = proposal.top_block();
-                        if !self.chain_info_provider.is_block_imported(&block) {
-                            debug!(target: "aleph-data-store", "Requesting a stale block {:?} after it has been missing for {:?} secs.", block, time_waiting.as_secs());
-                            self.block_requester
-                                .request_stale_block(block.hash, block.num);
-                        }
-                    }
+            if self.bump_proposal(&proposal) {
+                continue;
+            }
+            // `bump_proposal` returns false if the bump didn't make the proposal available, meaning that it is still pending
+            let time_waiting = match now.duration_since(first_occurrence) {
+                Ok(tw) if tw >= self.config.request_block_after => tw,
+                _ => continue,
+            };
+
+            let block = proposal.top_block();
+            if !self.chain_info_provider.is_block_imported(&block) {
+                debug!(target: "aleph-data-store", "Requesting a stale block {:?} after it has been missing for {:?} secs.", block, time_waiting.as_secs());
+                self.block_requester
+                    .request_stale_block(block.hash, block.num);
+                continue;
+            }
+            // The top block (thus the whole branch, in the honest case) has been imported. What's holding us
+            // must be that the parent of the base is not finalized.
```

Comment:
is this possible for an honest scenario?

Reply:
oh, it must have been a hopeless fork, right? if so I would encourage you to mention it, otherwise please correct me

Reply:
This is either because of a malicious proposal, or because we are not up-to-date with finalization (which is the main thing to deal with here). I've adjusted the comment to reflect that piece of information.